### PR TITLE
Refactor Makefile

### PR DESCRIPTION
**Pull request by mpoullet**
Goal is to ease cross-compilation:

* add PROTOC and PROTOC_PLUGINS_PATH variables
* add GOOGLEAPIS_ASSISTANT_PATH variable with API path and version
* remove references to /usr/local by using pkg-config
* remove redefinition of default variable CXX
* use default variable AR
* fix broken target $(GOOGLEAPIS_ASSISTANT_CCS:.cc=.h)
* add protobufs and clean to PHONY target
* add a HAS_PKG_CONFIG variable to avoid it if not available

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -24,37 +31,54 @@ GOOGLEAPIS_RPC_CCS = $(shell find $(GOOGLEAPIS_GENS_PATH)/google/rpc \
...
+                           $(GOOGLEAPIS_GENS_PATH)/$(GOOGLEAPIS_ASSISTANT_PATH)/embedded_assistant.grpc.pb.cc
 
 HOST_SYSTEM = $(shell uname | cut -f 1 -d_)
 SYSTEM ?= $(HOST_SYSTEM)
-CXX = g++
-CPPFLAGS += -I/usr/local/include -pthread -I$(GOOGLEAPIS_GENS_PATH) \
-	    -I$(GRPC_SRC_PATH) -I./src/
-CXXFLAGS += -std=c++11
+
+HAS_PKG_CONFIG ?= true
```
Comment (Fleker) [inline Makefile]: Will this ever be false?

File: Makefile [hunk 2]
```diff
@@ -24,37 +31,54 @@ GOOGLEAPIS_RPC_CCS = $(shell find $(GOOGLEAPIS_GENS_PATH)/google/rpc \
...
+ifeq ($(HAS_PKG_CONFIG),true)
+GRPC_GRPCPP_CFLAGS=`pkg-config --cflags grpc++ grpc`
+GRPC_GRPCPP_LDLAGS=`pkg-config --libs grpc++ grpc`
+ALSA_CFLAGS=`pkg-config --cflags alsa`
+ALSA_LDFLAGS=`pkg-config --libs alsa`
+else
+GRPC_GRPCPP_CFLAGS ?=
```
Comment (Fleker) [inline Makefile]: Would we ever not want grpc to be compiled? ALSA is Linux-only, but grpc is a core part of the sample.

File: Makefile [hunk 3]
```diff
@@ -24,37 +31,54 @@ GOOGLEAPIS_RPC_CCS = $(shell find $(GOOGLEAPIS_GENS_PATH)/google/rpc \
...
+CXXFLAGS += $(ALSA_CFLAGS)
+LDFLAGS += $(ALSA_LDFLAGS)
 endif
 
 .PHONY: all
 all: run_assistant
 
 googleapis.ar: $(GOOGLEAPIS_CCS:.cc=.o)
-	ar r $@ $?
+	$(AR) r $@ $?
```
Comment (Fleker) [inline Makefile]: Where is `$(AR)` defined?

File: Makefile (hunk 1 above)
Comment (mpoullet) [reply Makefile]: It is the same technique used by the main gRPC Makefile to ease cross-compilation in the case where pkg-config is not available. To quote it:

<pre>
#
# The steps for cross-compiling are as follows:
# First, clone and make install of grpc using the native compilers for the host.
# Also, install protoc (e.g., from a package like apt-get)
# Then clone a fresh grpc for the actual cross-compiled build
# Set the environment variable GRPC_CROSS_COMPILE to true
# Set CC, CXX, LD, LDXX, AR, and STRIP to the cross-compiling binaries
# Also set PROTOBUF_CONFIG_OPTS to indicate cross-compilation to protobuf (e.g.,
#  PROTOBUF_CONFIG_OPTS="--host=arm-linux --with-protoc=/usr/local/bin/protoc" )
# <b>Set HAS_PKG_CONFIG=false</b>
# To build tests, go to third_party/gflags and follow its ccmake instructions
# Make sure that you enable building shared libraries and set your prefix to
# something useful like /usr/local/cross
# You will also need to set GRPC_CROSS_LDOPTS and GRPC_CROSS_AROPTS to hold
# additional required arguments for LD and AR (examples below)
# Then you can do a make from the cross-compiling fresh clone!
#
</pre>

File: Makefile (hunk 2 above)
Comment (mpoullet) [reply Makefile]: If you don't have pkg-config, you need a way to pass the CFLAGS/LDFLAGS needed to use gRPC, e.g. in my actual build script:

<pre>
    <...>
    local GRPC_GRPCPP_CFLAGS="-pthread -I${TARGET_INSTALL_PATH}/include"
    local GRPC_GRPCPP_LDLAGS="-L${TARGET_INSTALL_PATH}/lib -lgpr -laddress_sorting -lgrpc -lgrpc++"
    <...>
    GRPC_SRC_PATH=${GRPC_SRC_PATH} \
    GOOGLEAPIS_SRC_PATH=${GOOGLEAPIS_SRC_PATH} \
    GOOGLEAPIS_GENS_PATH=${GOOGLEAPIS_GENS_PATH} \
    HAS_PKG_CONFIG=false \
    SYSTEM=${SYSTEM} \
    GRPC_GRPCPP_CFLAGS=${GRPC_GRPCPP_CFLAGS} \
    GRPC_GRPCPP_LDLAGS=${GRPC_GRPCPP_LDLAGS} \
    CPPFLAGS=${CPPFLAGS} \
    LDFLAGS=${LDFLAGS} \
    CXX=${CXX} \
    AR=${AR} \
    PROTOC=${PROTOC} \
    PROTOC_PLUGINS_PATH=${PROTOC_PLUGINS_PATH} \
    make -C "${PROJECT_PATH}/assistant-sdk-cpp" clean run_assistant
</pre>

It's a way to help people with custom build systems without changing anything for the other users.

File: Makefile (hunk 3 above)
Comment (mpoullet) [reply Makefile]: `$(AR)` is a pre-defined variable in GNU make, like `$(CXX)`, see [here](https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html). By hard coding `ar` like this, there is no chance to properly cross-compile the SDK by passing its value from outside the Makefile.

File: Makefile (hunk 1 above)
Comment (mpoullet) [reply Makefile]: I've changed this a bit to make it like 100% similar to gRPC.

**Comment by Fleker** [comment]
This looks good to me, thanks.

**Pull request merged**